### PR TITLE
feat: add formula step

### PR DESCRIPTION
**Pull request by raphaelvignes**
Add the formula step.

### CODE REVIEW

**Comment by raphaelvignes** [comment]
[(5) [Snowflake SQL VQB] - Formula Step](https://trello.com/c/KQD21n8J/1117-5-snowflake-sql-vqb-formula-step)

File: server/weaverbird/backends/sql_translator/steps/formula.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    sql_query_describer: SQLQueryDescriber = None,
+    sql_translate_pipeline: SQLPipelineTranslator = None,
+) -> SQLQuery:
+    """
+    Supported operators are:
+    - "+"
+    - "-"
+    - "*"
+    - "/" -> To avoid
+    - "%"
```
Comment (f-bb-toucantoco) [inline server/weaverbird/backends/sql_translator/steps/formula.py:31]: If operator "%" is supported, I lthink it is a good idea to write a test :)

File: server/weaverbird/backends/sql_translator/steps/formula.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        f"query.transformed_query: {query.transformed_query}\n"
+        f"query.metadata_manager.tables_metadata: {query.metadata_manager.tables_metadata}\n"
+        f"query.metadata_manager.query_metadata: {query.metadata_manager.query_metadata}\n"
+    )
+    transformed_query = (
+        f"""{query.transformed_query}, {query_name} AS"""
+        f""" (SELECT {complete_fields(query)}, {handle_zero_division(step.formula)} AS {step.new_column}"""
+        f""" FROM {query.query_name})"""
+    )
+    query.metadata_manager.add_column(column_name=step.new_column, column_type='float')
```
Comment (f-bb-toucantoco) [inline server/weaverbird/backends/sql_translator/steps/formula.py:51]: Why a float for the new column ? 
If you make 1+1, type is an integer, we avoid a step to cast the number

File: server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py [hunk 3]
```diff
@@ -185,3 +186,10 @@ def complete_fields(query: SQLQuery, columns=None) -> str:
         compiled_query = ', '.join(query_keys)
 
     return compiled_query
+
+
+def handle_zero_division(formula: str) -> str:
+    if '/' not in formula:
+        return formula
+    else:
+        return re.sub(r'(?<=/)\s*(\w+)|(?<=/)\s*(\"?.*\"?)', r' NULLIF(\1\2, 0)', formula)
```
Comment (f-bb-toucantoco) [inline server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py:195]: a little comment to explain the regex please :)

File: server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py [hunk 4]
```diff
@@ -185,3 +186,10 @@ def complete_fields(query: SQLQuery, columns=None) -> str:
         compiled_query = ', '.join(query_keys)
 
     return compiled_query
+
+
+def handle_zero_division(formula: str) -> str:
+    if '/' not in formula:
```
Comment (f-bb-toucantoco) [inline server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py:192]: Question: we could div by 0, but for mod it is possible?

**Review by f-bb-toucantoco — CHANGES_REQUESTED** [review]

File: server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py (hunk 4 above)
Comment (raphaelvignes) [reply server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py:192]: Not it isn't, I'll update that.

File: server/weaverbird/backends/sql_translator/steps/formula.py (hunk 2 above)
Comment (raphaelvignes) [reply server/weaverbird/backends/sql_translator/steps/formula.py:51]: As seen together, for now: 
- We don't handle string operations
- Numeric operation results are cast as float 
- I create a document we improvements needed for the future.

File: server/weaverbird/backends/sql_translator/steps/formula.py (hunk 1 above)
Comment (raphaelvignes) [reply server/weaverbird/backends/sql_translator/steps/formula.py:31]: Done

File: server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py (hunk 4 above)
Comment (raphaelvignes) [inline server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py:192]: With this logic,  we avoid Snowflake's Zero Division Error. Error values are replaced by NULL.
The user will be able to replace the NULL values by ZERO using the replace step.
Later on we can enhance this by handling operation priorities etc...
I'll document that (and other "enhancement topics") in a a dedicated page.

File: server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py (hunk 3 above)
Comment (raphaelvignes) [reply server/weaverbird/backends/sql_translator/steps/utils/query_transformation.py:195]: Added :)

**Review by f-bb-toucantoco — APPROVED** [review]

**Pull request merged**